### PR TITLE
IGNITE-17624 Sql. HashIndex support by optimizer

### PR DESCRIPTION
**Pull request by korlov42**

### CODE REVIEW

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ddl/DdlCommandHandler.java [hunk 1]
```diff
@@ -229,15 +232,20 @@ private static BiFunction<Object, Throwable, Boolean> handleTableModificationRes
     /** Handles create index command. */
     private CompletableFuture<Boolean> handleCreateIndex(CreateIndexCommand cmd) {
         Consumer<TableIndexChange> indexChanger = tableIndexChange -> {
-            // Only sorted idx for now.
-            createSortedIndexInternal(cmd, tableIndexChange.convert(SortedIndexChange.class));
+            if (cmd.type() == Type.SORTED) {
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ddl/DdlCommandHandler.java:235]: why not `switch` statement?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rel/AbstractIndexScan.java [hunk 2]
```diff
@@ -136,18 +144,24 @@ public RelOptCost computeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
 
             cost = 0;
 
-            if (lowerCondition() != null) {
-                double selectivity0 = mq.getSelectivity(this, RexUtil.composeConjunction(builder, lowerCondition()));
+            if (type == Type.HASH && lowerBound() != null) {
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rel/AbstractIndexScan.java:147]: could you please explain why lowerBound is importannt here?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/RexUtils.java [hunk 3]
```diff
@@ -328,6 +328,88 @@ public static IndexConditions buildSortedIndexConditions(
...
+        }
+
+        List<RexNode> searchCondition = new ArrayList<>();
+
+        Mappings.TargetMapping toTrimmedRowMapping = null;
+        if (requiredColumns != null) {
+            toTrimmedRowMapping = Commons.mapping(requiredColumns, rowType.getFieldCount());
+        }
+
+        for (int i = 0; i < indexedColumns.size(); i++) {
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/RexUtils.java:360]: why not simple for-each?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/RexUtils.java [hunk 4]
```diff
@@ -328,6 +328,88 @@ public static IndexConditions buildSortedIndexConditions(
...
+            String columnName = indexedColumns.get(i);
+
+            RelDataTypeField field = rowType.getField(columnName, true, false);
+
+            if (field == null) {
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/RexUtils.java:365]: which case you cover by the condition? I can't imaging situation when field will be absent

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ddl/DdlCommandHandler.java (hunk 1 above)
Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ddl/DdlCommandHandler.java]: fixed

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rel/AbstractIndexScan.java (hunk 2 above)
Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rel/AbstractIndexScan.java:150]: have added the comment to the code. Is it helpful?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/RexUtils.java (hunk 3 above)
Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/RexUtils.java]: I believe this is matter of taste. Anyway changed it

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rel/AbstractIndexScan.java (hunk 2 above)
Comment (ygerzhedovich) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rel/AbstractIndexScan.java:150]: yes, now it is clear. Thanks

**Review by ygerzhedovich — APPROVED** [review]

**Review by zstan — APPROVED** [review]

**Comment by zstan** [comment]
@korlov42 looks good

**Pull request merged**